 ## 5-2 Network Training:

We consider the following error function:

$~~~~~~~~~~~~~~~~~~~~~~~$    $ E(w)=\sum_{n=1}^{N} \{y(x_{n}, w)-t_{n}\}^2$

our goal is to minimize this function, such there no explicit solution to this problem , so this problem is an $optimization$ problem.

one approach widely used is getting the minimal (it may be local) through $gradient$ $~$ $descent$ algorithm.

## 5-2-4 Gradient Descent Optimization:


$gradient$ $~$ $descent$ is an itertive algorithm for optimization, it $n^{th}$ iteration  giveen by the following expression:

$~~~~~~~~~~~~~~~~~~~~~$ $~$ $w^{(n+1)}=w^{(n)}-\eta \nabla E(w^{(n)})$

where:

$\eta$ is called $ learning$ $~$ $rate$, it represent the pace of the training.

$\nabla E(w^{(n)})$ the derivative of the error function with respect the weights

$gradient$ $~$ $descent$ is an itertive algorithm , we need first need to initailize the weights(we prefer to initialize the weights randomly to avoid 
stabilities of some weights due the symetry...etc)

we consider the following example:

$~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~$   $f(w1, w2)=w_{1}^2+w_{1}+1+w_{2}^2+4w_{2}+3$

as we can the minimum of this function is $\hat w=(1, -2)$

we have:

$~~~~~~~~~~~~~~~~~~~~~~~~~~~~$ $\nabla f(w1, w2)=[2w_{1}-2, 2w_{2}+4]$

we choose:

$\eta=.01 $

$w_{0}$ a random vector.

number of iteration is 1000.

In [20]:
import numpy as np

def gradient1(w1):
    return 2*w1-2

def gradient2(w2):
    return 2*w2+4
w01, w02=np.random.random(), np.random.random()
n_iterations=1000
learning_rate=.01

for __ in range(n_iterations):
    w01=w01-learning_rate*gradient1(w01)
    w02=w02-learning_rate*gradient2(w02)
w=(w01, w02)
w

(0.9999999995847941, -1.9999999956960288)

as you can see $(0.9999999995847941, -1.9999999956960288)\approx (1, -2)$

## Error Backpropagation:

in the real world , when we want to train our $neural$ $~$ $networks$ we deal with multiple layers (more than input and ouput layer), so computing 
the $gradient$ of the error function rely on the chain formula.

let's be $E(w)$ the error function and $y(x_{n}, w)$ our model's prediction, if i want to compute $E(w)$ with respect the weight $w$
the chain formula give us:

$~~~~~~~~~~~~~~~~~~~~~~~~~~$ $\dfrac{dE(w)}{dw}=\dfrac{dE(w)}{dy_{n}}*\dfrac{dy_{n}}{dw}$

we applied this formula repaeditely untill we reatch the input layer.

when we obtain the $ gradient$ we applied the gradient descent algorithm.


for example check out the FeedForward network functions notebooks.